primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,399 @@
+	///
+	/// Only called if not expired just yet.
+	/// Note that the deadline is only passed to allow some optimizations
+	/// for some `L` types.
+	fn snooze(&self, _deadline: &Self) {
+		sp_io::offchain::sleep_until(offchain::timestamp().add(Duration::from_millis(
+			STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE,
+		)));
+	}
+}
+
+impl Lockable for Timestamp {
+	fn current() -> Self {
+		offchain::timestamp()
+	}
+
+	fn deadline() -> Self {
+		Self::current().add(Duration::from_millis(
+			STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS,
+		))
+	}
+
+	fn expired(&self, reference: &Self) -> bool {
+		<Self as Lockable>::current() > *reference
+	}
+
+	fn snooze(&self, deadline: &Self) {
+		let remainder: Duration = self.diff(&deadline);
+		// do not snooze the full duration, but instead snooze max 100ms
+		// it might get unlocked in another thread
+		// consider adding some additive jitter here
+		let snooze = core::cmp::min(remainder.millis(), STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE);
+		sp_io::offchain::sleep_until(self.add(Duration::from_millis(snooze)));
+	}
+}
+
+/// Lockable based on the current block number and a timestamp based deadline.
+#[derive(Debug, Copy, Clone, Encode, Decode)]
+pub struct BlockAndTime<B>
+where
+	B: BlockNumberTrait,
+{
+	/// The block number that has to be reached in order
+	/// for the lock to be considered stale.
+	pub block_number: B,
+	/// Additional timestamp based deadline which has to be
+	/// reached in order for the lock to become stale.
+	pub timestamp: Timestamp,
+}
+
+impl<B> Lockable for BlockAndTime<B>
+where
+	B: BlockNumberTrait,
+{
+	fn current() -> Self {
+		Self {
+			block_number: B::current(),
+			timestamp: offchain::timestamp(),
+		}
+	}
+
+	fn deadline() -> Self {
+		Self::current()
+	}
+
+	fn expired(&self, reference: &Self) -> bool {
+		let current = <Self as Lockable>::current();
+		current.timestamp > reference.timestamp && current.block_number > reference.block_number
+	}
+
+	fn snooze(&self, deadline: &Self) {
+		let remainder: Duration = self.timestamp.diff(&(deadline.timestamp));
+		let snooze = core::cmp::min(remainder.millis(), STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE);
+		sp_io::offchain::sleep_until(self.timestamp.add(Duration::from_millis(snooze)));
+	}
+}
+
+/// Storage based lock.
+///
+/// A lock that is persisted in the DB and provides a mutex behaviour
+/// with a defined safety expirey deadline based on a [`Lockable`](Self::Lockable)
+/// implementation.
+pub struct StorageLock<'a, L>
+where
+	L: Sized + Lockable,
+{
+	// A storage value ref which defines the DB entry representing the lock.
+	value_ref: StorageValueRef<'a>,
+	deadline: L,
+}
+
+impl<'a, L> StorageLock<'a, L>
+where
+	L: Lockable,
+{
+	/// Create a new storage lock with [default expiry duration](Self::STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS).
+	pub fn new<'k>(key: &'k [u8]) -> Self
+	where
+		'k: 'a,
+	{
+		Self {
+			value_ref: StorageValueRef::<'a>::persistent(key),
+			deadline: L::deadline(),
+		}
+	}
```

Comment:
```suggestion
	pub fn new(key: &'a [u8]) -> Self
	{
		Self {
			value_ref: StorageValueRef::<'a>::persistent(key),
			deadline: L::deadline(),
		}
	}
```

why do you need two lifetimes here if Self has only one?

Reply:
True